### PR TITLE
Started documenting interfaces

### PR DESCRIPTION
**Pull request by alex**
I couldn't really figure out where to start describing them, and also this example sucks. Could use some suggestions!

### CODE REVIEW

File: samples/test-siphash.bagel [hunk 1]
```diff
@@ -7,7 +7,7 @@ class TestSiphash:
     def test_from_paper(self):
         key = b"\x00\x01\x02\x03\x04\x05\x06\x07\08\x09\x0a\x0b\x0c\x0d\x0e\x0f"
         data = b"\x00\x01\x02\x03\x04\x05\x06\x07\08\x09\x0a\x0b\x0c\x0d\x0e"
-        assert siphash.siphash_2_4(key, data).assert_success() == 0xa129ca6149be45e5
+        assert siphash_2_4(key, data).assert_success() == 0xa129ca6149be45e5
```
Comment (dstufft) [inline samples/test-siphash.bagel:10]: I'm confused, is this related to interfaces somehow?

Comment (alex) [inline samples/test-siphash.bagel:10]: Not at all, let me move that over to master.

**Comment by dstufft** [comment]
How do we know that `TestPassed` and `TestFailure` implement the `TestResult` interface?

**Comment by alex** [comment]
They implement it by providing all the required methods. If they didn't,
and you tried to pass a `TestPassed` or `TestFailure` instance somewhere
that took a `TestResult` you'd get a compiler error letting you know what
methods were missing.

On Tue, May 13, 2014 at 10:21 AM, Donald Stufft notifications@github.comwrote:

> How do we know that TestPassed and TestFailure implement the TestResultinterface?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/19#issuecomment-42960990
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by dstufft** [comment]
So it's like uh, Compiler verified duck typing?

**Comment by alex** [comment]
Yeah!

On Tue, May 13, 2014 at 10:23 AM, Donald Stufft notifications@github.comwrote:

> So it's like uh, Compiler verified duck typing?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/19#issuecomment-42961288
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by dstufft** [comment]
Hmm. I'm not sure how I feel about that. It feels like it'll just introduce ambiguity which a static type system typically prevents.

**Comment by alex** [comment]
FWIW, this is basically exactly the way that Go interfaces work.

On Tue, May 13, 2014 at 10:28 AM, Donald Stufft notifications@github.comwrote:

> Hmm. I'm not sure how I feel about that. It feels like it'll just
> introduce ambiguity which a static type system typically prevents.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/19#issuecomment-42961928
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by dstufft** [comment]
Yea, it just feels implicit and magical and completely non intuitive.

**Comment by alex** [comment]
Do you prefer the java style where you declare that your class implements a
particular interface?

On Tue, May 13, 2014 at 10:31 AM, Donald Stufft notifications@github.comwrote:

> Yea, it just feels implicit and magical and completely non intuitive.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/19#issuecomment-42962388
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by dstufft** [comment]
I've never used Java (or Go really) but I think that would be better. Looking at this snippet of code I had no idea how to "connect the dots" from a TestResults to the classes that implement them.

Something like:

```
class TestPassed:
    implements TestResult

    def print_result(self, w: io.Writer):
        w.write(b"PASSED\n")
```

Makes it more obvious even to someone whose never seen the language before.

**Comment by alex** [comment]
@dstufft here's what I think the problem with that is: you don't always know what interfaces you're going to implement when you write a class.

For example:

Consider a memcache client library, it exposes a class like:

```
class MemcacheClient:
    def get(self, key: Bytes) -> Bytes:
        # ...
```

You're writing a web app, you're writing some tests, but you don't want to use memcached in testing, you just want to use an in memroy cache. With implicit interfaces, your project could just add:

```
interface class CacheClient:
    def get(self, key: Bytes) -> Bytes
```

and then use that throughout the code base. With explicit interfaces, either the memcache library would need to include an interface (which is stupid, what if you decide you want to use redis instead? Now it has to include yet another cache client interface?), or you'd have to write an adapter class that does nothing but proxy methods through so it can declare that it participates in an interface.

Do you find that persuasive?

**Comment by dstufft** [comment]
So why can't you just make it possible to declare something as implementing an interface outside of the class definition. zope.i and ABCs allow that and it seems like a reasonable way of solving that which doesn't involve a whole bunch of implicitness. Something isn't a CacheClient just because it has a get() method.

**Comment by alex** [comment]
So, it's totally true that something isn't a `CacheClient` just because it has a `get()` (just like a `Pizza` isn't a `Tire` just because it provides `diameter()`). I think the verbosity / comfort feel makes the tradeoff worth it though. Like you originally said: it's compiler checked duck typing.

**Comment by alex** [comment]
Ok, I sat down and wrote some real prose, let me know what you think.

File: docs/features/interfaces.rst [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+Consider a web application which needs a cache of some sort, in production you
+probably will use Memcached, but in local testing you might want to use a
+simple in-memory cache. In other languages you might implement a scheme like
+this an abstract base class and each implementation inheriting from it, but
+Bagel doesn't have subclassing.
+
+First, we write our interface::
+
+    interface class CacheClient
```
Comment (dstufft) [inline docs/features/interfaces.rst]: I assume interfaces do not have an implicit `__new__` and if you want that as part of your interface you'll need to `def` it here yes?

Comment (alex) [inline docs/features/interfaces.rst]: Interface shouldn't even define a constructor, for the same reason you
never see an `__init__` in a zope.interface or an ABC.

On Sun, May 18, 2014 at 11:15 AM, Donald Stufft notifications@github.comwrote:

> In docs/features/interfaces.rst:
> 
> > @@ -0,0 +1,53 @@
> > +Interfaces
> > +==========
> > +
> > +Interfaces are a way to write code in Bagel where you don't know precisely what
> > +types you will be dealing with.
> > +
> > +Consider a web application which needs a cache of some sort, in production you
> > +probably will use Memcached, but in local testing you might want to use a
> > +simple in-memory cache. In other languages you might implement a scheme like
> > +this an abstract base class and each implementation inheriting from it, but
> > +Bagel doesn't have subclassing.
> > +
> > +First, we write our interface::
> > +
> > -    interface class CacheClient
> 
> I assume interfaces do not have an implicit **new** and if you want that
> as part of your interface you'll need to def it here yes?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/19/files#r12775156
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (dstufft) [inline docs/features/interfaces.rst]: So how do you define an interface on how you create a class?

Comment (alex) [inline docs/features/interfaces.rst]: You don't. You need to explicitly create an instance of some class, and
then you can pass it somewhere where you need an instance of the interface.
Can you show an example of the kind of code you're thinking about?

On Sun, May 18, 2014 at 11:16 AM, Donald Stufft notifications@github.comwrote:

> In docs/features/interfaces.rst:
> 
> > @@ -0,0 +1,53 @@
> > +Interfaces
> > +==========
> > +
> > +Interfaces are a way to write code in Bagel where you don't know precisely what
> > +types you will be dealing with.
> > +
> > +Consider a web application which needs a cache of some sort, in production you
> > +probably will use Memcached, but in local testing you might want to use a
> > +simple in-memory cache. In other languages you might implement a scheme like
> > +this an abstract base class and each implementation inheriting from it, but
> > +Bagel doesn't have subclassing.
> > +
> > +First, we write our interface::
> > +
> > -    interface class CacheClient
> 
> So how do you define an interface on how you create a class?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/19/files#r12775430
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (dstufft) [inline docs/features/interfaces.rst]: Well for example, in Python packaging a `setup()` call can pass in a distclass which the `setup()` call will instantiate instead of the standard `Distribution` class that ships with distutils (or setuptools). So I figure that for anything that requires an instance of `Distribution` you'd instead have an interface for that, but that doesn't solve the problem of what if the constructor for the passed in class doesn't match the constructor for the original `Distribution` class?

Comment (alex) [inline docs/features/interfaces.rst]: In bagel you can't pass around class objects like that, and instantiate
them later. The bagel way to do that would be for setup to take a Function
type, which has some specific arguments, and returns an instance of the
interface.

On Sun, May 18, 2014 at 11:21 AM, Donald Stufft notifications@github.comwrote:

> In docs/features/interfaces.rst:
> 
> > @@ -0,0 +1,53 @@
> > +Interfaces
> > +==========
> > +
> > +Interfaces are a way to write code in Bagel where you don't know precisely what
> > +types you will be dealing with.
> > +
> > +Consider a web application which needs a cache of some sort, in production you
> > +probably will use Memcached, but in local testing you might want to use a
> > +simple in-memory cache. In other languages you might implement a scheme like
> > +this an abstract base class and each implementation inheriting from it, but
> > +Bagel doesn't have subclassing.
> > +
> > +First, we write our interface::
> > +
> > -    interface class CacheClient
> 
> Well for example, in Python packaging a setup() call can pass in a
> distclass which the setup() call will instantiate instead of the standard
> Distribution class that ships with distutils (or setuptools). So I figure
> that for anything that requires an instance of Distribution you'd instead
> have an interface for that, but that doesn't solve the problem of what if
> the constructor for the passed in class doesn't match the constructor for
> the original Distribution class?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/19/files#r12775456
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (dstufft) [inline docs/features/interfaces.rst]: Not sure how I feel about that, but I guess if that's how it's done then it makes sense not to interface on the constructor.

Comment (alex) [inline docs/features/interfaces.rst]: There's a reasonable argument that you should be able to easily convert a
class to a function; but that's kind of orthagonal. I've never seen an
interface system that allowed/where it was typical to put constructors in
the interface /cc @glyph @dreid

On Sun, May 18, 2014 at 11:26 AM, Donald Stufft notifications@github.comwrote:

> In docs/features/interfaces.rst:
> 
> > @@ -0,0 +1,53 @@
> > +Interfaces
> > +==========
> > +
> > +Interfaces are a way to write code in Bagel where you don't know precisely what
> > +types you will be dealing with.
> > +
> > +Consider a web application which needs a cache of some sort, in production you
> > +probably will use Memcached, but in local testing you might want to use a
> > +simple in-memory cache. In other languages you might implement a scheme like
> > +this an abstract base class and each implementation inheriting from it, but
> > +Bagel doesn't have subclassing.
> > +
> > +First, we write our interface::
> > +
> > -    interface class CacheClient
> 
> Not sure how I feel about that, but I guess if that's how it's done then
> it makes sense not to interface on the constructor.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/19/files#r12775469
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (dstufft) [inline docs/features/interfaces.rst]: I dunno, I've never used an interface system really in any meaningful capacity :) It just seems silly to require a function to create a class when we already have the constructor which creates a class. I mean what you're specifying you'd essentially have two interfaces, the "instance" interface and the "constructor" interface. I don't really seem a whole lot of benefit in having them separate.

Comment (glyph) [inline docs/features/interfaces.rst]: The fact that `Foo` and `Foo()` mean such radically different things in Python is often super confusing to people learning the language.  Even I forget or add too many parens sometimes, and the error messages when one does that are less than helpful.

If I really wanted a grim sort of revenge on the Python community, I'd say that every class should have an attribute called "`factory`".  But, I'm not bitter, so, how about `Foo.create`?

Comment (glyph) [inline docs/features/interfaces.rst]: Also: why `interface class foo` and not just `interface foo`?

Comment (alex) [inline docs/features/interfaces.rst]: It provides a sort of consistency, any time you write `class Foo` it means
you can use that name as a type (`enum class Foo`, `interface class Foo`,
`class Foo`). (Does this make any sense? It felt natural to me, but I might
be loopy).

On Sun, May 18, 2014 at 1:37 PM, Glyph notifications@github.com wrote:

> In docs/features/interfaces.rst:
> 
> > @@ -0,0 +1,53 @@
> > +Interfaces
> > +==========
> > +
> > +Interfaces are a way to write code in Bagel where you don't know precisely what
> > +types you will be dealing with.
> > +
> > +Consider a web application which needs a cache of some sort, in production you
> > +probably will use Memcached, but in local testing you might want to use a
> > +simple in-memory cache. In other languages you might implement a scheme like
> > +this an abstract base class and each implementation inheriting from it, but
> > +Bagel doesn't have subclassing.
> > +
> > +First, we write our interface::
> > +
> > -    interface class CacheClient
> 
> Also: why interface class foo and not just interface foo?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/19/files#r12776162
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (dstufft) [inline docs/features/interfaces.rst]: Oh I didn't even notice that. I agree that `interface class Foo` seems like extra typing for little benefit.

Comment (alex) [inline docs/features/interfaces.rst]: @glyph With the suggest of maybe having `Foo.create`, should `Foo()` even be the standard constructo if we accept that? Why not always just use `Foo.create()`?

Comment (glyph) [inline docs/features/interfaces.rst]: Actually, I do think that's a better method.  Smalltalk and ObjC both go with `Object new.` as the default go-to constructor.  This also avoids the bizarre `__new__`-is-a-classmethod-but-it-doesn't-say-so circularity in Python.  All constructors are just classmethods, and the only "special" one is the one that does instance allocation.  (`new` is a convention that does `self ← class alloc. self init.`

That leaves the parens-vs-no-parens dichotomy for a separate issue…

File: docs/features/interfaces.rst [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        def get(self, key: Bytes) -> Option<Bytes>
+        def set(self, key: Bytes, value: Bytes)
+
+This says that to implement the ``CacheClient`` interface, a class must define
+``get`` and ``set`` methods, which those signatures.
```
Comment (dreid) [inline docs/features/interfaces.rst]: s/which/with

File: docs/features/interfaces.rst (hunk 2 above)
Comment (dreid) [inline docs/features/interfaces.rst]: This should end in a colon?

Comment (alex) [inline docs/features/interfaces.rst]: Fixed.

On Mon, May 19, 2014 at 9:12 AM, David Reid notifications@github.comwrote:

> In docs/features/interfaces.rst:
> 
> > @@ -0,0 +1,53 @@
> > +Interfaces
> > +==========
> > +
> > +Interfaces are a way to write code in Bagel where you don't know precisely what
> > +types you will be dealing with.
> > +
> > +Consider a web application which needs a cache of some sort, in production you
> > +probably will use Memcached, but in local testing you might want to use a
> > +simple in-memory cache. In other languages you might implement a scheme like
> > +this an abstract base class and each implementation inheriting from it, but
> > +Bagel doesn't have subclassing.
> > +
> > +First, we write our interface::
> > +
> > -    interface class CacheClient
> 
> This should end in a colon?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/19/files#r12802378
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084
